### DISCUSSION TITLE
Filename Schema - put DATE after TITLE

### DISCUSSION THREAD

**Discussion by ryanwwest** [novoid/Memacs#124] [Ideas]
Hi, I was wondering if in the filename schema in https://github.com/novoid/Memacs/blob/master/docs/memacs_filenametimestamps.org, if it would be possible to put the timestamp after the 'title' of the file? For example, for `2010-12-31T23.59_Cookie_recipies.pdf` instead have the file be recognized by `Cookie_recipies_2010-12-31T23.59.pdf`?

My reasoning is that I tried a similar schema from https://karl-voit.at/managing-digital-photographs/ of DATE-TITLE-TAGS.EXTENSION in the past but having the date first made it difficult to identify each file (mainly using the title), hard to select files in the terminal without fuzzy finders (where you have to type the start of the file first), and more frustrating to use in any other program that isn't set up for them. I love the idea in general and appreciate that DATE first creates a nice universal sort order, but I'd prefer to put TITLE first and am not sure if everything can be accomplished in the same way that way. 

For example, can Org links still refer just to the (unique) timestamp, not necessarily the whole filename, if the filename doesn't start with the date?

**Comment by novoid** [comment]
Wow, my first "Github discussion" I guess. ;-)

I'm sorry, DATE-TITLE-TAGS.EXTENSION is my file name convention and I did not implement it that way so that its order can be changed easily.

However, I'm curious. What do you mean by "identify each file" or "hard to select files in the terminal without fuzzy finders"? I'm identifying files and selecting files all the time without any issue. Maybe I can help you there?

Related to Org links, you might want to read https://karl-voit.at/2022/02/10/lfile/ where I explain my current method to link files. It's not using Memacs any more. The only workflow I'm using myself that makes use of the Memacs filenametimestamp module is [adding images to lazyblorg](https://github.com/novoid/lazyblorg/wiki/Images) besides the usual "visualize files in the Memacs agenda" of course.

**Reply by ryanwwest** [reply]
> However, I'm curious. What do you mean by "identify each file" or "hard to select files in the terminal without fuzzy finders"

My biggest use case is bash in the terminal, using tab completion which starts by typing out the beginning of the file name and pressing <tab> which then displays only the remaining files in that directory that match what you typed out to not have to always type out the entire filename (but only from the beginning of the filename). bash and zsh have the same default tab completion behavior that I think most beginner users use, and I don't often customize terminals heavily since I have to ssh into many different machines so I usually stick to defaults. Thus, if I have 20 files in a directory created this last year that all start with 2023-... then that's at least 6 keystrokes before I type the first more unique character in order to get tab completion. Something like [fzf](https://www.redhat.com/sysadmin/fzf-linux-fuzzy-finder) would probably fix it but since it's not default I haven't gotten in the habit of using it.

Other than that, when scanning lists of files in for example Mac Finder / Windows File Explorer, it's faster for me to identify files by title and slightly a chore in my head to jump past the dates to do this. But I'm being seriously pedantic—all dates are the same length anyway.

It's not a problem if Memacs doesn't currently support this. I'd love to hear how you get around the above if it's even a problem. My question was more, does this alternative schema make sense, and are there any technical constraints that would make this more difficult (even outside the scope of Memacs for general file organization)? I'd imagine the script is relying on some sort of regex with ^<DATE> so only looking at the beginning and I was curious how hard the change might be and if it would impact performance significantly.

> Related to Org links, you might want to read https://karl-voit.at/2022/02/10/lfile/ where I explain my current method to link files. It's not using Memacs any more. The only workflow I'm using myself that makes use of the Memacs filenametimestamp module is [adding images to lazyblorg](https://github.com/novoid/lazyblorg/wiki/Images) besides the usual "visualize files in the Memacs agenda" of course.

Thanks, I'll check it out.

By the way, I just realized you were Karl Voit - I recently DM'd you on Reddit but never made the connection, and have been following all your accounts as if you were two people. I really appreciate all your thoughts and contributions, your interests majorly overlap with mine!

**Reply by ryanwwest** [reply]
> Wow, my first "Github discussion" I guess. ;-)

It used to be issues for everything but if discussions are enabled I try to keep questions/theoretical stuff out to keep issue count down :) Of course it creates multiple places for sometimes similar discussions so maybe there are pros/cons.

**Reply by novoid** [reply]
ad me as a combination of Karl Voit, novoid, n0v0id (Twitter), publicvoit (homepage, Mastodon):

Unfortunately, I did not follow a consistent pattern with my aliases. :-(

Meanwhile, I try to use at least my real name in addition to my aliases so that people are able to link my account to my person.

ad shell UX:

I'm using more or less standard zsh in that regard, so we should both have the same UX. I often double click on the datestamp (I almost never use year-only date-stamps) and paste using the middle mouse in order to get a particular date into the current prompt line. As long as there aren't many files from the same day, that works fine. Of course, it also works fine for time-stamps as there are almost no files with the same time-stamp in my case. In case there are multiple files from the same day, multiple TABs do switch to completion mode suggesting one after another. That works file for me so far.

When I want to get a file from today, I use my [AutoKey](https://github.com/autokey/autokey) snippet `$$dd` which always inserts the day in form of "2023-09-26". So my workflow would be: "ls -la " + $$dd + TAB (or *) and get the appropriate result.  

If you're only the Memacs module, you probably should be fine with a fork + adapting the regex. Adapting the line `filename_timestamp_match = DATETIME_REGEX.match(file)` and its regex might be enough already.

However, I'm using many tools around https://karl-voit.at/managing-digital-photographs/ which follow my filename convention.

**Reply by ryanwwest** [reply]
> When I want to get a file from today, I use my [AutoKey](https://github.com/autokey/autokey) snippet $$dd which always inserts the day in form of "2023-09-26". So my workflow would be: "ls -la " + $$dd + TAB (or *) and get the appropriate result.

I do the same thing with the same ISO8601 formatted dates but it's alt/option-D across all systems. It's very valuable.

One reason why I asked about the flexible dates in filenames is because I want to design something that allows even more flexibility with dates and my data. I'd eventually like to build something that can read through markdown and Org files and search for YYYY-Q1, YYYY-MM and YYYY-MM-DD timestamps (optionally with additional time) at a heading, block/paragraph, and bullet point level and auto-include them into a sort of 'daily' journal unlike any other 'daily notes' system I've heard of to date. Meaning, if you added new thoughts to 5 notes files in a day, and 3 of them were important enough to warrant remembering by adding a date with our date-insert-shortcuts, then this tool would separately find each of the 3 files' headings(+content) / full paragraphs / bullet points (and sub-bullets) that listed that date and list them in a daily log as things that you edited, with hyperlinks to jump you to the source and maybe make it editable in-place. So in other words a sort of auto-journal where you don't have to think about actually journaling besides adding a timestamp if you're doing more than fixing typos or deleting stuff (which an alternative, git commits, might pick up as noise). 

Is this a sort of thing that might integrate well as a new Memacs module (or overlap with an existing one or other tool)? I'm not familiar enough with Memacs - perhaps my idea would fit better with [this](https://beepb00p.xyz/myinfra_files/myinfra.svg) (though they seem pretty related).

**Reply by novoid** [reply]
Well, it seems to be the case that you move the Memacs part where I'm using the Org agenda to the file level.

Memacs: open up the agenda, switch to the date you're interested in, activate the archive layer of your agenda in case you're using archive files for Memacs results and you've got your timeline view on your data.

You idea: open up a heading related to a date and see all the date-related linked files.

Valid approach but I prefer the standard agenda method. Furthermore, I would have no idea what to do with YYYY-Q1 files and day-based headings but that's an issue you need to resolve. ;-)

For now, I'll close this discussion. Feel free to post below if you have another thought related to that.

**Reply by ryanwwest** [reply]
Thanks for your thoughts. The Memacs agenda you describe where you switch to the date of interest and see what comes up is actually what I want too. I'll need to look up more deeply how it's displayed. It might be nice to jump to that date in the agenda view from any date of the SAME YYYY-MM-DD in an org file but that's not first priority. The biggest thing I want to figure out is finer granularity with which parts of the file are displayed in the agenda view as described above. I have some large files that have thoughts across many days and months so getting only the specific sections I wrote on a particular day would be nice (and editable from agenda view nicer).

**Comment by nbehrnd** [comment]
@ryanwwest Because of `cookierecipes` in the string of the file name, I initially thought about wanting to easy copy & past/yank (or move, delete) a bunch of files based on their name, rather than by the time stamp (`date2name`).

For one, *where suitable*, I use a file archive in similar shape as illustrated by the tree on Karl's blog post [My Folder Hierarchy](https://karl-voit.at/folder-hierarchy/).  Second, for simple navigation within my `~/.filetags_tagfilter/`, I started to use the cross-platform [vifm](https://vifm.info/) where a search *by name* (initiated by `/`) quickly narrows the selection if any of the string entered manually matches some portion of the file name.  As shown below, the date/time stamps by `date2name` typically are not an obstacle to rapidly identify the files in question.   For the purpose of the demonstration below, I replaced the default two-pane (Norton Commander like) display by a single column view:


![ezgif com-optimize](https://github.com/novoid/Memacs/assets/19709084/a5300498-37f2-4313-9beb-5ee09576640f)

**Reply by ryanwwest** [reply]
Cool, I love vim-based tools yet had somehow never heard of vifm. Partial filename matching like that is great. I'll check it out!

**Reply by nbehrnd** [reply]
@ryanwwest  I just noticed (frankly, once set up, it was forgotten because it worked ever since) one has to explicitly request the partial filename matching in vifm.  In my case, it likely is a combination of adjusting `set suggestions`, `set ignorecase`, `set smartcase` which is easiest documented by a share of the configuration file in the zip archive below.  Altogether with calling a color style file (fetched from the program's project [page](https://vifm.info/colorschemes.shtml), it yields a display as shown [elsewhere](https://fortran-lang.discourse.group/t/modern-fortran-logo/2075/30).

[example_vifm.zip](https://github.com/novoid/Memacs/files/12801438/example_vifm.zip)
